lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -2143,51 +2143,261 @@ namespace Js
+    // Specifically, parts of numbers can have sub-parts, but those sub-parts can't themselves have sub-parts.
+    // Thus, it returns a linear representation of a tree, where the implicit root node represents the entire
+    // formatted string and each child represents a specific part of the string. Per ICU documentation, parts
+    // will *never* overlapped -- any given part can only be a strict subset of exactly one part, a strict
+    // superset of one or more parts, or no relationships at all.
+    // The tree we want to form looks as follows:
+    //
+    // [1] [2] [    3    ] [4] [    5    ]
+    //              |             |   |
+    //             [6]           [7] [8]
+    //
+    // We want to convert the above tree back into an array of parts that looks as follows:
+    //
+    // [1], [2], [3.1], [6], [3.2], [4], [5.1], [7], [5.2], [8], [5.3]
+    //
+    // In more concrete words, we want to assign the entire formatted string to some number of parts,
+    // and each character should get the most specific part possible. We can accomplish this by
+    // keeping an auxiliary array with the same size as the formatted string that, at each index,
+    // holds the part for that character. For the following call pattern:
+    //
+    // 1) Add Integer field [0, 9)
+    // 2) Add grouping separator field [5, 6)
+    // 3) Add grouping separator field [1, 2)
+    //
+    // The array will transform as follows:
+    //
+    // 0) [0, 0, 0, 0, 0, 0, 0, 0, 0]
+    // 1) [int, int, int, int, int, int, int, int, int]
+    // 2) [int, int, int, int, int, group, int, int, int]
+    // 3) [int, group, int, int, int, group, int, int, int]
+    //
+    // Which lines up with the expected resulting part array:
+    //
+    // [
+    //    { type: integer, value: "1" },
+    //    { type: group, value: "," },
+    //    { type: integer, value: "000" },
+    //    { type: group, value: "," },
+    //    { type: integer, value: "000" },
+    // ]
+    //
+    // We can (and should) also handle cases where we get grouping separators before the integer,
+    // or when we get grouping separator, integer, grouping separator
+    class NumberFormatPartsBuilder
+    {
+    private:
+        double num;
+        const char16 *formatted;
+        const charcount_t formattedLength;
+        ScriptContext *sc;
+        UNumberFormatFields *fields;
+
+        static const UNumberFormatFields UnsetField = static_cast<UNumberFormatFields>(0xFFFFFFFF);
+
+        JavascriptString *GetPartTypeString(UNumberFormatFields field)
+        {
+            JavascriptLibrary *library = sc->GetLibrary();
+
+            if (field == UnsetField)
+            {
+                return library->GetIntlLiteralPartString();
+            }
+
+            switch (field)
+            {
+            case UNUM_INTEGER_FIELD:
+            {
+                if (JavascriptNumber::IsNan(num))
+                {
+                    return JavascriptString::NewCopySz(_u("nan"), sc);
```

Comment:
Honestly, I didn't want to type out the numbers for each length to do NewWithBuffer, and regardless I planned on moving these to JavascriptLibrary's string cache before merging. This was mostly to prove that it would work.